### PR TITLE
Track Endpoint [OTP-1189]

### PR DESCRIPTION
**Pull request by binh-dam-ibigroup**
### Checklist

- [x] Appropriate branch selected _(all PRs must first be merged to `dev` before they can be merged to `master`)_
- [x] Any modified or new methods or classes have helpful JavaDoc and code is thoroughly commented
- [na] The description lists all applicable issues this PR seeks to resolve
- [na] The description lists any configuration setting(s) that differ from the default settings
- [x] All tests and CI builds passing

### Description

This PR adds a new `track` endpoint that can be used to indifferently start or update a tracked trip (roughly forwarding to the same code as `starttracking` and `updatetracking` respectively). Under the hood, payload and response types are streamlined.

### CODE REVIEW

File: src/main/java/org/opentripplanner/middleware/triptracker/ManageTripTracking.java [hunk 1]
```diff
@@ -224,12 +213,49 @@ public static TrackedJourney getOngoingTrackedJourney(String tripId) {
...
         try {
-            return getPOJOFromRequestBody(request, payloadClass);
+            return getPOJOFromRequestBody(request, GeneralPayload.class);
         } catch (JsonProcessingException e) {
             logMessageAndHalt(request, HttpStatus.BAD_REQUEST_400, "Error parsing JSON tracking payload.", e);
             return null;
         }
     }
+
+    private static class TripData {
```
Comment (binh-dam-ibigroup) [inline src/main/java/org/opentripplanner/middleware/triptracker/ManageTripTracking.java]: @br648 Would it make sense to merge this class with `GeneralPayload`?

File: src/main/java/org/opentripplanner/middleware/triptracker/response/TrackingResponse.java [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+
+import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
+import com.fasterxml.jackson.annotation.JsonInclude;
+
+/**
+ * The response provided to the caller when starting or updating tracking of a monitored trip.
+ */
+@JsonIgnoreProperties(ignoreUnknown = true)
+@JsonInclude(JsonInclude.Include.NON_NULL)
+public class TrackingResponse {
```
Comment (br648) [inline src/main/java/org/opentripplanner/middleware/triptracker/response/TrackingResponse.java:11]: This is much cleaner. I should have done this first time around.

File: src/main/java/org/opentripplanner/middleware/controllers/api/TrackedTripController.java [hunk 3]
```diff
@@ -47,14 +47,20 @@ public void bind(final SparkSwagger restApi) {
                     .withDescription("Initiates the tracking of a monitored trip.")
                     .withProduces(JSON_ONLY)
                     .withRequestType(StartTrackingPayload.class)
-                    .withResponseType(StartTrackingResponse.class),
+                    .withResponseType(TrackingResponse.class),
                 (request, response) -> ManageTripTracking.startTracking(request), JsonUtils::toJson)
             .post(path("/updatetracking")
```
Comment (br648) [inline src/main/java/org/opentripplanner/middleware/controllers/api/TrackedTripController.java:52]: Do we even need the `/starttracking` and `/updatetracking` end points now? Seems that we could get away with just the `/track` end point? Would that cause issues with the mobile guys?

File: src/main/java/org/opentripplanner/middleware/triptracker/ManageTripTracking.java (hunk 1 above)
Comment (br648) [reply src/main/java/org/opentripplanner/middleware/triptracker/ManageTripTracking.java]: @binh-dam-ibigroup I'd be inclined to keep them separate. `GeneralPayload` for all items that could be passed to the end points and `TripData` for serverside tie-up between monitored trip and tracking. Perhaps move `TripData` into its own class?

File: src/main/java/org/opentripplanner/middleware/triptracker/ManageTripTracking.java [hunk 4]
```diff
@@ -224,12 +213,49 @@ public static TrackedJourney getOngoingTrackedJourney(String tripId) {
...
+        public final List<TrackingLocation> locations;
+
+        public TripData(MonitoredTrip trip, TrackedJourney journey, List<TrackingLocation> locations) {
+            this.trip = trip;
+            this.journey = journey;
+            this.locations = locations;
+        }
+    }
+
+    private static TripData getTripAndJourneyForUser(Request request) {
```
Comment (br648) [inline src/main/java/org/opentripplanner/middleware/triptracker/ManageTripTracking.java]: Nit. Method comment. Does every method need a comment?!

**Review by br648 — COMMENTED** [review]
A few comments, but on the whole this is good.

File: src/main/java/org/opentripplanner/middleware/controllers/api/TrackedTripController.java (hunk 3 above)
Comment (binh-dam-ibigroup) [reply src/main/java/org/opentripplanner/middleware/controllers/api/TrackedTripController.java:52]: Eventually, after the mobile app folks transitions to the `track` endpoint.

File: src/main/java/org/opentripplanner/middleware/triptracker/ManageTripTracking.java (hunk 4 above)
Comment (binh-dam-ibigroup) [reply src/main/java/org/opentripplanner/middleware/triptracker/ManageTripTracking.java]: I added JavaDocs in https://github.com/ibi-group/otp-middleware/pull/231/commits/6f547b46cfe61f7d9bdd752338e1d9f86de57aa8.

File: src/main/java/org/opentripplanner/middleware/triptracker/ManageTripTracking.java (hunk 1 above)
Comment (binh-dam-ibigroup) [reply src/main/java/org/opentripplanner/middleware/triptracker/ManageTripTracking.java]: I moved that and related methods in https://github.com/ibi-group/otp-middleware/pull/231/commits/2d58332f48ea829fa08b552e722684eb690b214c and https://github.com/ibi-group/otp-middleware/pull/231/commits/6f547b46cfe61f7d9bdd752338e1d9f86de57aa8

**Review by br648 — APPROVED** [review]
Good enhancement. Approved!

**Review by br648 — APPROVED** [review]
All good with me.

**Review by JymDyerIBI — APPROVED** [review]
👍  Looks good. I like the single endpoint.

**Pull request merged**